### PR TITLE
24-builder-prefixes

### PR DESCRIPTION
**Pull request by DarjaLeonova**

### CODE REVIEW

File: builder.go [hunk 1]
```diff
@@ -317,7 +318,7 @@ func (e *Expression) build(spacing string) string {
 		for _, o := range e.function.options {
 			s += " " + o.key + spacing + "=" + spacing
 
-			if v, ok := o.operand.(string); ok && len(v) > 0 && v[0] == '$' { // recognized as variable if string starts with $
+			if v, ok := o.operand.(string); ok && len(v) > 0 {
```
Comment (klavinsjanis) [inline builder.go:321]: It might fail on quoted literals after removal of - `&& v[0] == '$'`

Comment (janishorsts) [reply builder.go:321]: Also, the option value is not optional. E.g., an empty string should be `host = ||`.

File: builder.go [hunk 2]
```diff
@@ -175,7 +176,7 @@ func (b *Builder) Local(v string, expr *Expression) *Builder {
 		return b
 	}
 
-	b.locals = append(b.locals, local{variable: variable(v), expr: expr})
+	b.locals = append(b.locals, local{variable: variable(varSymbol + v), expr: expr})
```
Comment (janishorsts) [inline builder.go:179]: I would prefer to prefix '$' only when serializing/building to MF2 syntax.

File: builder.go [hunk 3]
```diff
@@ -361,27 +358,45 @@ type FuncOption struct {
...
 }
 
-func Option(key string, operand any) FuncOption { return FuncOption{key: key, operand: operand} }
-
 func Func(name string, option ...FuncOption) *Expression {
 	return Expr().Func(name, option...)
 }
 
+func VarOption(name, varName string) FuncOption {
+	return FuncOption{key: name, operand: varSymbol + varName}
```
Comment (janishorsts) [inline builder.go:366]: We use `variable` type in the `func (e *Expression) Var(name string)` but not here. Why?

It should be consistent - we cannot use `string` in one place and `variable` in another for the same type.

```suggestion
	return FuncOption{key: name, operand: variable(varName)}
```

File: builder.go [hunk 4]
```diff
@@ -338,20 +339,16 @@ func (e *Expression) Literal(v any) *Expression {
...
+	if len(name) == 0 {
 		panic("variable name cannot be empty")
 	}
 
-	if v[0] != '$' {
-		panic(fmt.Sprintf("variable must start with $: %s", v))
-	}
-
-	e.operand = variable(v)
+	e.operand = variable(varSymbol + name)
```
Comment (VladislavsPerkanuks) [inline builder.go:351]: same as @janishorsts mentioned. `$` symbol is used only for serialization, so we can skip it here

**Comment by klavinsjanis** [comment]
Would be good to see tests with the examples mentioned in the previous comments.

File: builder.go (hunk 1 above)
Comment (DarjaLeonova) [reply builder.go]: We don't need to write the dollar symbol before the variable anymore, so why will it fail?

**Review by klavinsjanis — APPROVED** [review]

File: builder.go (hunk 2 above)
Comment (VladislavsPerkanuks) [inline builder.go:179]: i thought we quit on adding prefixes on variables, and literals
```suggestion
	b.locals = append(b.locals, local{variable: variable(v), expr: expr})
```

File: builder.go [hunk 5]
```diff
@@ -338,20 +339,16 @@ func (e *Expression) Literal(v any) *Expression {
...
 
-func Var(s string) *Expression {
-	return Expr().Var(s)
+func Var(name string) *Expression {
+	return Expr().Var(name)
 }
 
-func (e *Expression) Var(v string) *Expression {
-	if len(v) == 0 {
+func (e *Expression) Var(name string) *Expression {
```
Comment (VladislavsPerkanuks) [inline builder.go:346]: please add checking if variable is valid. according to [abnf](https://github.com/unicode-org/message-format-wg/blob/7c00820a0462679eba696181c45bfadb43d2eedd/spec/message.abnf#L88-L95)

e.g. 
calling Var with argument "$var" should error/panic

File: builder.go [hunk 6]
```diff
@@ -361,33 +358,51 @@ type FuncOption struct {
 	key     string
 }
 
-func Option(key string, operand any) FuncOption { return FuncOption{key: key, operand: operand} }
+func VarOption(name, varName string) FuncOption {
+	return FuncOption{key: name, operand: variable(varName)}
```
Comment (VladislavsPerkanuks) [inline builder.go:362]: same 
add checking if variable is valid. according to [abnf](https://github.com/unicode-org/message-format-wg/blob/7c00820a0462679eba696181c45bfadb43d2eedd/spec/message.abnf#L88-L95)

File: builder.go (hunk 2 above)
Comment (DarjaLeonova) [reply builder.go:179]: Yes, you are right, forgot to remove it ;)

File: builder.go [hunk 7]
```diff
@@ -46,7 +48,7 @@ func (b *Builder) Build() (string, error) {
 	}
 
 	for _, v := range b.locals {
-		s += ".local" + b.spacing + string(v.variable) + b.spacing + "=" + b.spacing + v.expr.build(b.spacing) + b.newline
+		s += ".local" + b.spacing + varSymbol + string(v.variable) + b.spacing + "=" + b.spacing + v.expr.build(b.spacing) + b.newline //nolint:lll
```
Comment (janishorsts) [inline builder.go:51]: It's ok to break this line

```suggestion
		s += ".local" + b.spacing + varSymbol + string(v.variable) + b.spacing + "=" +
		b.spacing + v.expr.build(b.spacing) + b.newline
```

File: builder.go [hunk 8]
```diff
@@ -544,3 +564,13 @@ func coalesce[T comparable](l ...T) T {
 
 	return c
 }
+
+// validateVarName checks whether the first rune is a valid starting character
+// for a variable name according to the specified criteria.
+func validateVarName(varName string) {
```
Comment (janishorsts) [inline builder.go:570]: Remove the validation - I'd rather have no validation than an incomplete one, it validates only the first character.

Two issues:
* panics, interrupts the build flow - we store error in the Builder struct
* incomplete validation.

**Review by VladislavsPerkanuks — APPROVED** [review]

**Pull request merged**